### PR TITLE
Don't reuse RandomState seeds

### PR DESCRIPTION
**Pull request by arthurprs**
cc #36481

### CODE REVIEW

**Comment by alexcrichton** [comment]
@rfcbot fcp merge 

Looks good to me, thanks @arthurprs! Want to run it by @rust-lang/libs just to be super sure though we're on board with this solution to #36481

**Comment by sfackler** [comment]
I would not describe this as a solution to #36481. Performance is still unnacceptable with non-default hashers. It does work around the issue in the common case, and we should land it though.

**Comment by arthurprs** [comment]
Do you think I need to reword the comment mentioning #36481?

File: src/libstd/collections/hash/map.rs [hunk 1]
```diff
@@ -2049,24 +2050,21 @@ impl RandomState {
...
-        thread_local!(static KEYS: (u64, u64) = {
+        // Later in #36481 it was discovered that exposing a deterministic
+        // iteration order allows a form of DOS attack. To counter that we
+        // increment one of the seeds on every RandomState creation, giving
+        // every corresponding HashMap a different iteration order.
+        thread_local!(static KEYS: UnsafeCell<(u64, u64)> = {
             let r = rand::OsRng::new();
             let mut r = r.expect("failed to create an OS RNG");
-            (r.gen(), r.gen())
+            UnsafeCell::new((r.gen(), r.gen()))
```
Comment (alexcrichton) [inline src/libstd/collections/hash/map.rs]: I believe this `UnsafeCell` can be `Cell`, right?

**Comment by alexcrichton** [comment]
The in-code comment is fine to me, but I changed `Solves #...` in the description to `cc #...`

File: src/libstd/collections/hash/map.rs (hunk 1 above)
Comment (arthurprs) [inline src/libstd/collections/hash/map.rs]: True, I'll change it.

**Comment by brson** [comment]
I agree this is a good and efficient interim solution. To be clear, how much confidence do we have that addition here is _actually_ a mathematically valid way to ensure that merging two hashmaps won't collide? My understanding is that it is clearly better than the status quo, that we suspect that _in practice_ it will prevent collisions, but there's not strictly any mathematically-proven basis for _knowing_ it will be effective (i.e. some clever person may find a way to generate collisions by incrementing the initialization value).

**Comment by brson** [comment]
[Here are the benchmarks](https://github.com/rust-lang/rust/issues/36481#issuecomment-256730627) comparing this approach of incrementing this value to completely regenerating it. This approach is 2ns vs 6-15ns.

**Comment by arthurprs** [comment]
The fix relies on the backing hash function being high-quality, as in "Every input bit should affect every output bit about ~50% of the time" (seed is also an input).

**Comment by pczarn** [comment]
@arthurprs The seed is not an input. The seed describes a choice of a function from a family of functions.

Regardless, the seed and the input are mixed in a similar way. Perhaps the solution is effective. I don't know cryptography. I can only read what cryptographers have explained.

**Comment by Manishearth** [comment]
Shouldn't `Clone` also reinitialize the random state? We should never be copying the random state afaict.

**Comment by funny-falcon** [comment]
But what if change order of iteration just will fix original issue?

We can use just simplest lcg generator for walking order:

```
  // Pseudo code for iterator
  if (hash.size == 0) return
  i := 0
  do {
    if (slot_occupied(hash, i))
       yield hash.key[i]
    i = (i *5 + 1) & hash.mask
  } until (i != 0) /* returned to first position, so we walked all hash */
```

**Comment by funny-falcon** [comment]
@sfackler points out that given output of iteration one may construct new input values in desired order to DoS the application.

Ok, lcg is very tunable, so we can show different order on every iteration:

```
  // Pseudo code for iterator
  if (hash.size == 0) return
  ptis = ptis * 5 + 1 //per_thread_iteration_seed
  delta := ptis * 2 + 1
  mult := ptis * 4 + 1
  i := 0
  do {
    if (slot_occupied(hash, i))
       yield hash.key[i]
    i = (i * mult + delta) & hash.mask
   } until (i != 0)
```

**Comment by arthurprs** [comment]
@pczarn good point. I'll open the pr if nobody else does.

@funny-falcon the problem with that approach is that the performance hit for iteration is possibly large. It should be easily measured though.

**Comment by funny-falcon** [comment]
@arthurprs you right: cache misses on big hashtables will destroy performance. So, my suggestion is not acceptable.

**Comment by sacundim** [comment]
Quoting the [SipHash paper](https://131002.net/siphash/siphash.pdf), pp. 5-6 (my boldface):

> Note that the standard PRF and MAC security goals allow the attacker access to the output of SipHash on messages chosen adaptively by the attacker. However, they do not allow access to any “leaked” information such as bits of the key or the internal state. They also do not allow **“related keys”**, “known keys”, “chosen keys”, etc.

In the patch attached to this pull request, the `RandomState::new()` function increments the second word of the key on each call.  This means that `RandomState`s created in the same thread have **related keys**, which violates the preconditions of SipHash's security claims.  Does anybody have an argument that SipHash achieves its security goal under these conditions?

One candidate solution that does not fall afoul of the random keys precondition would be:

1. Use the same seed for many maps.
2. Associate each map with *unique ID* that differs between all maps that share the same seed.
3. Instead of hashing inputs verbatim, prefix them with the map's unique ID.

With thread-local seeds, the unique ID can be an additional `u64` counter.  Since SipHash supports incremental computation, one simple optimization is to precompute, cache and reuse the hasher's state after absorbing the prefixed counter.  A further optimization would be to defer this step until the first time the map needs to hash an input so that creating empty maps does not pay the cost of that.

**Comment by funny-falcon** [comment]
I think, it is a time to ask Jean-Phillipe Aumasson @veorq : does SipHash allows incrementing key, or will it fail? even though attacker never sees whole output of hashsum (it can only guess low bits, and even those bits are not exact cause of collision resolution)?

**Comment by veorq** [comment]
As noted above we didn't make claims in the case of "related keys", but I haven't seen attack in the related-key model. 

TL;DR: I think it's ok.

Details: IIUC the scenario here is the following:

- at time t1 you compute SipHash( `key`, _)
- at time t2 you'll compute SipHash( `key + 1`, _)
- at time t2 you'll compute SipHash( `key + 2`, _)
- etc. 

The attacker can choose what messages are hashed. We can assume that the attacker sees the hashes of the messages with `key + i` for different `i`'s and wants to predit SipHash(`key + j`, m) for some `j>i`. The attacker will only succeed if there's some relation between hashes of different messages (standard PRF security) or if there's some relation between hashes of different or identical messages with different keys (related-key PRF security).

Given the previously published cryptanalysis results on SipHash, I'm confident that SipHash will remain secure under related keys and as used in Rust.
